Batch processing

takojunior opened the pull request:
The PR is to enable pattern mining on batches of sequences instead of being only able to run on entire set. The changes included in this PR include:

- Enable mining on batches of sequences. Mining tasks can run in parallel
- Enable backend mining task to run on a child process which improves memory efficiency
- Enable aggregation of patterns mined from batches as final results
- Require that the input integer sequences must contain positive integers
- Fixes on the docs and readme files

bkleyn commented on .github/workflows/ci.yml:
```diff
@@ -14,7 +14,7 @@ jobs:
       matrix:
         python-version: ['3.6', '3.7', '3.8', '3.9', '3.10']
-        os: [ubuntu-latest, macos-latest, windows-latest]
+        os: [ubuntu-20.04, macos-latest, windows-latest]
```
Why fix this? Does it not work on latest?

bkleyn replied:
Got it. Might be worth dropping 3.6 at some point.

skadio replied:
if dropping 3.6 gets us latests versions, something to consider in future

takojunior replied:
It shows that as of 2021-12-23, 3.6 has already reached the [end of life phase](https://peps.python.org/pep-0494/#lifespan). So it doesn't seem to be supported any more in ubuntu releases. I think it makes sense to drop 3.6.

bkleyn commented on sequential/utils.py:
```diff
@@ -232,6 +234,10 @@ def validate_sequences(sequences: List[list]):
+    if not isinstance(sequences[0][0], str):
+        min_int = get_min_value(sequences)
+        check_true(min_int > 0, ValueError("Integers that are representing items cannot contain 0. "
+                                           "The integers must be positive integers."))
```
Wouldn't it be sufficient to say: "Integers must be be greater than 0"?

takojunior replied:
Re: the analysis of speed up, for example in the unit [test](https://github.com/fidelity/seq2pat/blob/batch_processing/tests/test_seq2pat_batch.py#L265) on test data with not constraints. 

-  Single thread with no batching: 1.4s
-  Single thread with batching, batch_size_10000 : 6.9s
-  8 threads with batching, batch_size_10000 : 2.1s

So there is an effect of speed up. Given a smaller set, ~60k sequences, there seems no need to go batching.

bkleyn commented on sequential/seq2pat.py:
```diff
@@ -423,27 +464,26 @@ def get_patterns(self, min_frequency: Num) -> List[list]:
 
-        # Cython implementor object with input parameters set
-        self._cython_imp = self._get_cython_imp(min_frequency)
+            # Check min_frequency conditions
```
Can you add a few comments on why this is added? It does not look only specific to batch?

takojunior replied:
Right, this is not specific to batch. Given a set of sequences, either a batch or the entire set of sequences, we would need to validate if min_frequency parameter is set correctly. There exist edge cases when min_frequency is not properly set, e.g. 0, the program will fail.

bkleyn commented on tests/test_seq2pat.py:
```diff
@@ -1301,7 +1302,7 @@ def test_repeated_calls_seq2pat(self):
 
         # Sequential pattern finder
-        seq2pat = Seq2Pat(sequences)
+        seq2pat = Seq2Pat(sequences, batch_size=2)
```
What is the default batch_size? Is behavior same as before?

takojunior replied:
`batch_size=None` by default, so it is running on entire set on a single thread as before. When it is set, the mining will be enabled to run on `batch_size` of sequences in parallel.

skadio replied:
n_jobs default is set to 2, no? Is this still same as before?

skadio replied:
Btw, not sure why this test is changing?

takojunior replied:
This test has been covered in test_seq2pat_batch and the original test should be the same as before.  This file should keep unchanged.

bkleyn reviewed (commented):
Looks good to me at high-level. Added a few clarifying comments/questions.

Have you done any analysis to evaluate speed-up and/or improved scalability?

takojunior commented on sequential/seq2pat.py:
```diff
@@ -458,6 +498,89 @@ def get_patterns(self, min_frequency: Num) -> List[list]:
+                                                                                    (chunk_ind + 1) * self.batch_size])
+                batch_seq2pat.add_constraint(new_constraint)
+
+        # Reduce min_frequency to be min_frequency * min_frequency_df, 0 < min_frequency_df < 1
```
Will add more comments here. So I am applying a lower min_frequency threshold in mining patterns from individual batch. The new threshold is calculated by applying a factor with value in 0-1 to min_frequency, `min_frequency * min_frequency_df`, where min_frequency is a percentage. The result will be smaller than min_frequency.

skadio replied:
cannot this be calculated from the given original min_freq and the batch_size.. If I am looking at %30 frequency at the whole data, and now I want 2 batches, would not that be %15 in each batch?

takojunior replied:
You mean the row counts, not the percentage right? For row counts, yes we can calculate but I think percentage would be easier. In terms of percentage, for each batch, it will still be 30%, then the sum of two batches will become 30% of entire set.

dorukkilitcioglu replied:
Ah, had the same comment above..

skadio commented on README.md:
```diff
@@ -22,6 +22,10 @@ Seq2Pat is developed as a joint collaboration between Fidelity Investments
+
+We present two quick start examples as follows. Notice that the input sequences in each example have items 
+represented as strings, while Seq2Pat also allows input to be sequence of integers and the integers must be positive.
+
```
possible rephrase;

We present examples for constraint-based sequential pattern mining and dichotomic pattern mining. Sequences can be represented as strings or positive integers.

skadio commented on sequential/seq2pat.py:
```diff
@@ -283,9 +293,27 @@ class Seq2Pat:
         sequences but no constraints are used to run the mining efficiently and practically.
         Power users can choose to drop this constraint by setting it to be None or increase the maximum span
         as the system has resources to support.
+    batch_size: Optional[int]
```
What's the relationship between batch_size and n_jobs. Do you split data into S sizes, that gives B bactches, and we execute on B batches with N jobs? 

Also, is this gonna be reproducible?

skadio replied:
Need more commentary on how batching works. Is this batched sequentially, or randomly etc. and how is the results aggregated

takojunior replied:
Yes, the entire set is split into B batches with each batch having `batch_size` sequences. Then I map these batches to N jobs.

Before the entire set is split, I shuffle the sequences, trying to make the sequences uniformly distributed. Then batches are split sequentially. There is a random seed used for data shuffle, to make it reproducible.

The patterns are aggregated from mining the batches by calculating the sum of their occurrences.

takojunior replied:
The `min_frequency_df` parameter is used to reduce min_frequency threshold for each batch. It should be set small enough to capture the patterns those should satisfy the min_frequency threshold while running on the entire set. 

Theoretically there is a chance that the batching results will be different from non-batching results. But a small `min_frequency_df` parameter will make the chance to be minimal and thus we have the same results as running on entire set.

dorukkilitcioglu replied:
Just trying to understand.. If `min_frequency` was `4` and we set batch size to be half the whole dataset, couldn't we use `min_frequency` of `2` for each batch to get a similar result?

skadio replied:
I am trying to understand this, but still confused. In particular, not sure why we need both of batch_size and n_jobs together. And on top, we also need min_freq_df, which I am not sure either. 

I have a feeling that we should be able to simplify this

takojunior replied:
> Just trying to understand.. If `min_frequency` was `4` and we set batch size to be half the whole dataset, couldn't we use `min_frequency` of `2` for each batch to get a similar result?

Yes, that is the main idea for batching process to work. Reducing `min_frequency` (as a percentage) is the way used to apply a lower threshold for each batch. However, for some patterns that can pass this threshold on entire set, those cannot pass the new threshold on each batch. We need to figure out the way to apply an even lower threshold for each batch.

takojunior replied:
> I am trying to understand this, but still confused. In particular, not sure why we need both of batch_size and n_jobs together. And on top, we also need min_freq_df, which I am not sure either.
> 
> I have a feeling that we should be able to simplify this

For using `batch_size` and `n_jobs`, the idea is to control the data size loaded in for one mining task, so this will use less memory for each task, and leave it to users to decide how many cpus can be used. 

For `min_freq_df` parameter, it defines a degree that we want to reduce the threshold for each batch, since not all patterns that should be included in the results as it is done on entire set, would pass the original threshold on a batch, a subset of data.

skadio commented on sequential/utils.py:
```diff
@@ -243,3 +248,121 @@ def validate_max_span(max_span: Union[int, None]):
+            attr_to_cs[attr][cs] = new_constraint
+
+    return sequences, attr_to_cs
+
```
empty lines

skadio commented on tests/test_seq2pat_batch.py:
```diff
@@ -0,0 +1,284 @@
+
+        seq2pat = Seq2Pat(sequences=[["A", "C", "B", "A", "D"],
+                                     ["C", "B", "A"],
+                                     ["C", "A", "C", "D"]], batch_size=2)
```
so how does this work? It uses first 2 rows as batch 1 and the remaining 1 row as batch 2? IF there were 6 rows, we would do batches of 2 x 3? And if there were 7 rows, we would a batch of 3 + 4?

takojunior replied:
So if 6 rows, I think yes, it will go 2 x 3. Then if 7 rows, it will go 2 x 3, then 1 x 1, making it 4 batches.

skadio commented on tests/test_seq2pat_batch.py:
```diff
@@ -0,0 +1,284 @@
+        # API and Cython object test. Replicates command line:
+        # ./MPP -file input.txt -thr 0.01 -out
+        # input on Main.cpp and verifies output with data captured from original implementation
+        # Unconstrained call. Significantly different and larger results.
```
what does significantly different/larger result mean?

takojunior replied:
The comments should be updated. Here it means running `./MPP -file input.txt -thr 0.01 -out` yields significantly different result comparing to the default command line [example](https://github.com/fidelity/seq2pat/blob/batch_processing/sequential/seq2pat.py#L682) where a few constraints are enforced. I shall update accordingly.

skadio commented on sequential/utils.py:
```diff
@@ -42,12 +44,12 @@ def read_data(source: str, is_scientific: bool = False) -> List[list]:
 
 
-def string_to_int(mapping: map, items: List[List[int]]) -> List[list]:
+def string_to_int(mapping: map, items: List[List[str]]) -> List[List[int]]:
```
thanks for fixing this! :D

skadio commented on sequential/utils.py:
```diff
@@ -243,3 +248,121 @@ def validate_max_span(max_span: Union[int, None]):
+
+def validate_min_frequency_with_batch(num_rows, batch_size, min_frequency):
+    """
+    Validate min_frequency when Seq2Pat runs on batches
```
can you add a word on what's the expecttation for a valid or invalid min_freq for the batch case?

takojunior replied:
> can you add a word on what's the expecttation for a valid or invalid min_freq for the batch case?

A valid min_freq would be a float that makes `batch_size * min_freq` to generate anrow count as an integer greater than 0. 

User can define `batch_size`, while there would be one batch (the last one when we split sequentially) containing less sequences than `batch_size`. So validation step also checks if `min_freq` works with the last batch.

skadio commented on sequential/utils.py:
```diff
@@ -243,3 +248,121 @@ def validate_max_span(max_span: Union[int, None]):
+    return sort_pattern(aggregated_patterns)
+
+
+def shuffle_data(sequences, attr_to_cs, seed):
```
what's attr_to_cs. A quick comment in pydoc of what this function would help

skadio commented on sequential/seq2pat.py:
```diff
@@ -283,9 +293,27 @@ class Seq2Pat:
+        when mining tasks are applied among batches in parallel. If -1 all CPUs are used.
+    seed: int
+        Random seed to make sequences uniformly distributed among batches.
+    min_frequency_df: float
```
I am not following what this parameter does.. also the name "df" is confusing, making the user thing it requires a dataframe.

skadio replied:
why do we need this parameter.. cannot we "auto" calculate sth using the other min_freq parameter and batch_size?

takojunior replied:
The parameter tries to make `min_freq` to be a lower value for a batch, so the results from mining a batch will contain more patterns, to avoid a pattern that slightly falls below threshold gets dropped from batch, and thus dropped in the final results. I would think over how we can simply it.

skadio commented on sequential/seq2pat.py:
```diff
@@ -423,27 +464,28 @@ def get_patterns(self, min_frequency: Num) -> List[list]:
         # Check num_rows
         check_true(self._num_rows >= 1, ValueError("Sequences should not be empty."))
 
-        # Check min_frequency conditions
```
thanks for lifting this into a utility

skadio commented on sequential/seq2pat.py:
```diff
@@ -423,27 +464,28 @@ def get_patterns(self, min_frequency: Num) -> List[list]:
 
-        # Frequent mining
-        patterns = self._cython_imp.mine()
+            # Call Cython backend in a child process, write results to queue
```
are we gettign same results when nothing is used in here?

takojunior replied:
Yes, we get the same results. The existing unit tests can get passed.

skadio commented on sequential/seq2pat.py:
```diff
@@ -458,6 +500,89 @@ def get_patterns(self, min_frequency: Num) -> List[list]:
 
+    def _get_patterns_batch(self, min_frequency) -> List[list]:
+        """
+        Get patterns from each batch with a relaxed min_frequency threshold, then aggregate patters among batches.
```
say more on how exactly "aggregate". typo "patters"

skadio reviewed (changes requested):
I didn't do a super detailed review, but needs a bit more explanations. 

This reminds me somethign we overlooked. We are making a considerable Public API change. As we did before, it would have been more efficient to do an API Review even before attempting the particular implementation. I am currently hang up on the API change, per review comments

dorukkilitcioglu commented on tests/test_seq2pat_batch.py:
```diff
@@ -0,0 +1,284 @@
+        # Patterns
+        patterns = seq2pat.get_patterns(min_frequency=0.8)
+
+        seq2pat = Seq2Pat(sequences=[["A", "C", "B", "A", "D"],
```
While this does work, based on the comments above, you could get a different result, right?

takojunior replied:
Yes, the idea is to make `min_frequency` to be proportionally smaller on each batch. But the occurrence of a pattern on a batch can be slightly larger or smaller than the desired threshold. If it is slightly smaller, we would still want to return it from the batch, so one way for doing this is to reduce the threshold to some degree. Then there will be an aggregation of patterns to make sure the finally returned patterns meet desired occurrence threshold.

dorukkilitcioglu commented on tests/test_seq2pat_batch.py:
```diff
@@ -0,0 +1,284 @@
+
+        self.assertEqual(patterns, batch_patterns)
+
+    def test_input_no_upper_constraint_batch(self):
```
Can you tell a bit on why we're trying to replicate that output?

takojunior replied:
This is to check Seq2Pat can give the same results as we run the C++ package. So we know the Cython part is running correctly.

dorukkilitcioglu reviewed (commented):
Looks good overall. I also need some clarification on the batch size vs the discount factor, but otherwise I'm onboard with the changes.

takojunior commented:
> Looks good to me at high-level. Added a few clarifying comments/questions.
> 
> Have you done any analysis to evaluate speed-up and/or improved scalability?



Re: the analysis of speed up, for example in the unit [test](https://github.com/fidelity/seq2pat/blob/batch_processing/tests/test_seq2pat_batch.py#L265) on test data, ~50k sequences, without constraints.

Single thread with no batching: 1.4s
Single thread with batching, batch_size_10000 : 6.9s
8 threads with batching, batch_size_10000 : 2.1s
So there is an effect of speed up. Given a smaller set, there seems no need to go batching.

skadio commented on .github/workflows/ci.yml:
```diff
@@ -13,7 +13,7 @@ jobs:
     strategy:
       matrix:
-        python-version: ['3.6', '3.7', '3.8', '3.9', '3.10']
+        python-version: ['3.7', '3.8', '3.9', '3.10']
```
cool! Can we make sure the requirement for Python 3.7+ matches what we mention in Readme, Docs etc.

takojunior commented on sequential/seq2pat.py:
```diff
@@ -151,6 +155,12 @@ def values(self):
         """
         return self._values
 
+    def set_values(self, values):
```
Any advices for better practices? This is called in `_mining_batch()` method, which is for internal use then, it might be better to make it to be internal as well.

takojunior commented:
> I am done with my first pass.
> 
> To be direct, I am not comfortable with a PR that came a little too late in the year and just weeks before the AAAI Tutorial. This PR changes BOTH the API and the Behavior.. and with a tutorial ahead of us, for this PR to fly, it should have been crystal clear --which is unfortunately not the case.
> 
> High-level comments:
> 
> * We should have had a quick API Design Review beforehand (note to ourselves)
> * What's the functionality of this feature? I might have missed it, but what happens when one runs this? What are the runtime benefits of batching? What are experimental gains/observations? Any analysis on batch size/runtime/dataset? There is a quick comment on pydocs about setting batch-size. what analysis is that based on? More importantly, apart from runtime, what happens to Patterns? Do we get the roughly same set of patterns? Are patterns now completely different when batched? How does pattern change a function of bath_size()? What should user expect on pattern changes?

Agree that this PR includes some major updates to the API and mining behaviors when it goes batching. I will send another note to the reviewers and facilitate discussions as the follow-ups to resolve comments. Given the AAAI tutorial will be in a few weeks, the PR would not be ready to be pushed through. 

I think it might be better to withdraw the PR and take the comments as follow-up actions. I will resubmit the PR once I have reviewers' questions and comments addressed more clearly.

takojunior commented:
Hi All,

I would like to reopen this PR for the consideration of adding batch processing capability. Since the last review,  I made a few updates to the PR:
- Updated naming of parameter `min_frequency_df` to `discount_factor`
- Revised comments in unit tests for clarity, mostly in `test_seq2pat_batch.py` for those to test if results from original c++ implementation and seq2pat are the same
-  Moved shuffle_data method to be internal to Seq2Pat, since it's only used in batching model. Users might not necessarily need to see it.
- Moved set_attribute to be internal to Attribute, since that is also only used in batching. It is not necessarily to be seen.
- Added validations of the parameters used in batching, e.g. `batch_size`, `n_jobs`, `discount_factor`
- Added an example notebook `batch_processing.ipynb` to show the usage of batch mode on `sample_data`.

For the analysis on batch_size/discount_factor/runtime, we addressed the questions in last PR with more experiments. Here is a summary of analysis:
- We tested on a data set with approximately 100k sequences. The results show that when `batch_size` increases , e.g. from 10000 to 100000, we observe an increase in runtime, while the mined patterns are all the same as mining on the entire set using single thread. When `batch_size=10000`, we get the most runtime benefit when it is compared to running on entire set.
- On the same 100k sequences, we set `batch_size=10000` and change `discount_factor` from 0.1 to 1.0. We observe that the runtime decreases as `discount_factor` increases. Only when `discount_factor=1.0`, the batching mode will miss some patterns compared to running on entire set. We would recommend `discount_factor=0.2` by default for the robustness in results, at the expenses of runtime.
- In an even larger test on ~1M sequences, we set `batch_size=10000`, `discount_factor=0.8`,  `n_jobs=8`. Batch mode saves 60% of the runtime compared to running on entire set, while the resulted patterns from the two processes are the same.
- When data size is small, e.g. a few thousands sequences, there is no benefit to run batch mode. Thus we would recommend to use the batch mode only when data has at least hundreds of thousands sequences for gaining the runtime benefit.

Let me know if you have more questions.

skadio commented on sequential/seq2pat.py:
```diff
@@ -283,12 +294,36 @@ class Seq2Pat:
+        entire set. A mining task will run on each batch with a reduced minimum row count (min_frequency) threshold.
+        Please refer to description of discount_factor parameter for how min_frequency is reduced.
+        Resulted patterns will be aggregated from the mining results of each batch by calculating the sum of
+        the occurrences. Finally the original minimum row count threshold is applied to the patterns after aggregation.
```
for batch_size parameter, add a comment about what's the default behaviour.

skadio replied:
BTW, I would immediately copy the Experimental Results Summary here.. Taht would greatly help a power user reading the docs. 

* Some experimental analysis for batch_size vs. discount_factor vs. runtime tested on a data set with 100k sequences. 
* The results show that when batch_size increases , e.g. from 10000 to 100000, we observe an increase in runtime, while the mined patterns are all the same as mining on the entire set using single thread. When batch_size=10000, we get the most runtime benefit when it is compared to running on entire set.
* On the same 100k sequences, we set batch_size=10000 and change discount_factor from 0.1 to 1.0. We observe that the runtime decreases as discount_factor increases. Only when discount_factor=1.0, the batching mode will miss some patterns compared to running on entire set. We would recommend discount_factor=0.2 by default for the robustness in results, at the expenses of runtime.
* In an even larger test on ~1M sequences, we set batch_size=10000, discount_factor=0.8, n_jobs=8. Batch mode saves 60% of the runtime compared to running on entire set, while the resulted patterns from the two processes are the same.
* When data size is small, e.g., a few thousand sequences, there is no benefit to run batch mode. Thus, we would recommend using the batch mode only when data has at least hundreds of thousands of sequences for gaining the runtime benefit.

skadio commented on sequential/seq2pat.py:
```diff
@@ -283,12 +294,36 @@ class Seq2Pat:
+        on batches in parallel. If -1 all CPUs are used. If -2, all CPUs but one are used.
+    seed: int
+        Random seed to make sequences uniformly distributed among batches.
+    discount_factor: float
```
should not we move this parameter right after batch_size where it is relevant?

skadio commented on sequential/seq2pat.py:
```diff
@@ -283,12 +294,36 @@ class Seq2Pat:
+        Random seed to make sequences uniformly distributed among batches.
+    discount_factor: float
+        A discount factor is used to reduce the minimum row count (min_frequency) threshold when Seq2Pat is applied
+        on a batch. When min_frequency is a integer, mining task can only be run on the entire set.
```
aN integer

skadio replied:
Not sure why can't we do batching with Integer min_freq.. cannot we find float version of a given min_freq_integer by looking at min_freq_integer and num_total_rows?

takojunior replied:
My immediate thought is to explicitly require it to be a float, then there would be less conversions or rounding between integer and ratio thresholds. But since the default seq2pat usage should support integer threshold, it might be better to allow integer in batch mode. I can add the automated conversion to it.

skadio commented on sequential/seq2pat.py:
```diff
@@ -283,12 +294,36 @@ class Seq2Pat:
+        A discount factor is used to reduce the minimum row count (min_frequency) threshold when Seq2Pat is applied
+        on a batch. When min_frequency is a integer, mining task can only be run on the entire set.
+        When min_frequency is a float, mining can be run on batches, with new threshold being defined by
+        max(min_frequency * discount_factor, 1.0/num_rows). num_rows is the number of sequences of one batch.
```
isn't num_rows == batch_size? IF so, why not use batch_size instead of introducing num_rows?

takojunior replied:
I am thinking that there would be one batch having number of sequences less than defined batch_size. So I introduce num_rows, which could be more accurate. But this may not be critical or of user's interests. I can use batch_size directly.

skadio commented on sequential/seq2pat.py:
```diff
@@ -458,6 +505,89 @@ def get_patterns(self, min_frequency: Num) -> List[list]:
+            delayed(self._mining_batch)(i, sequences, attr_to_cs, min_frequency)
+            for i in range(0, num_chunks))
+
+        # Aggregate patterns from each chunk over the counts and apply min_frequency threshold
```
this is resolved now

skadio commented on sequential/utils.py:
```diff
@@ -243,3 +248,110 @@ def validate_max_span(max_span: Union[int, None]):
+    Transform counter of patterns to list
+    """
+    results = []
+    for key, value in patterns_counter.items():
```
same q as before

takojunior replied:
I missed this question before. So I tried an implementation of vectorizing, e.g. transform the dictionary to an array and then do the array operations instead of for-loop, but it seems the two methods have similar performance in tests, and the for-loop is slightly faster. Maybe it is because I still need to transform back to list of list as the final format in returns, where I need a loop. So I might keep this implementation for now until we find a better solution.

skadio replied:
yep, I see

skadio commented on sequential/utils.py:
```diff
@@ -243,3 +248,110 @@ def validate_max_span(max_span: Union[int, None]):
+def aggregate_patterns(batch_patterns, min_row_count):
+    counter = collections.Counter()
+    for patterns in batch_patterns:
+        counter.update(list_to_counter(patterns))
```
same q as before

takojunior replied:
Here it is intended to use the counter to add up the frequencies for each pattern first, so we go from list to counter for each batch and aggregate, and then go back from final counter to list as the result and it satisfies min_freq. This is the solution for the aggregation so far.

skadio replied:
gotcha --thanks for the explanation!

skadio commented on sequential/seq2pat.py:
```diff
@@ -283,12 +294,36 @@ class Seq2Pat:
     """
 
-    def __init__(self, sequences: List[list], max_span: Optional[int] = 10):
+    def __init__(self, sequences: List[list], max_span: Optional[int] = 10, batch_size=None,
```
shall we consider re-ordering parameters so that batch stuff is together 

batch_size, discoun_factor, n_jobs, seed

skadio replied:
So the Default behavior is None? Can you say more, why? I thought this was to make "naive" users not to run into memory issues, similar to fixing max_span=10 so that it runs fast. Shall we consider enforcing batch by default, or is it too dramatic? 

Btw, what happens if Batch_size is larger than num_rows in the original data? I assume we handle that inside (and should have a unit test for that edge case, if not already)

takojunior replied:
Done with the reordering of parameters. It makes sense to put batch parameters together.

Regarding the default behavior for batch, I assume the dataset might not be in a size of a few millions in many cases, so running on entire set won't be an issue then and it can be faster. For example, in one analysis, the data is 10000, if I set batch_size=10000 as default, it is slower than not using batch due to the introduced operations, e.g. shuffle, discount factor... The runtime benefit will be obvious only when we deal with large dataset. So I think we can make default behavior to be single thread for now.

And yes, if batch_size is larger than num_rows, we can still get right results. I have a test to cover this.

skadio reviewed (approved):
LGTM. See some of comments before the final merge

skadio commented on sequential/seq2pat.py:
```diff
@@ -283,12 +294,36 @@ class Seq2Pat:
         as the system has resources to support.
+    batch_size: Optional[int]
+        When batch_size is set, program runs Seq2Pat on batches of sequences instead of on the entire set of
+        sequences for improving scalability. Each batch contains `batch_size` sequences as a random sample of
```
please bold **random** .. This is important behavior

skadio commented on sequential/seq2pat.py:
```diff
@@ -151,6 +156,12 @@ def values(self):
         """
         return self._values
 
+    def _set_values(self, values):
```
This function was not needed before, but now necessarily. Can you add a comment inside to say why is this needed and where it is used (sth like to re-set the constraints within batch etc.. )

takojunior replied:
Yes, this function is needed when we split the data into batches. 

In each batch, we need to set the correct attributes and the constraints. Since the upper and lower bound are set on the entire set, it would be easier to copy these constraints but only reset the attribute values in a batch. 

I am prevented to set the values directly, so I add the set_value function. I will add a comment inside.

skadio replied:
exactly! Commenting the above in the code would be great

takojunior commented:
Hi All, I believe all comments have been resolved. A few recent commits make these updates:
- Enable dynamically apply batch processing. For regular users, when batch_size is not set but dataset has >500k sequences, we automatically set batch_size=10000 to ease the task and improve user experience.

- Allow integer min_frequency threshold in batch processing. The value will be converted to a ratio for the algorithm to run.

- Add an internal random state to seq2pat, which implements the shuffle process.

- Clarify the notes for set_value function, batch_size attribute in seq2pat, attach experimental analysis to comments.

- Update docs.

I plan to merge and have the new release either today or tomorrow. 

Thank you all again for the constructive comments!
